### ISSUE TITLE
non-generic CachePolicy does not work wrapped in a non-generic PolicyWrap

### ISSUE THREAD

**Issue by TDabasinskas** [App-vNext/Polly#341]
Hello,

I have the following `WrapPolicy` defined:
```csharp
var cachePolicy = Policy.CacheAsync(new MemoryCacheProvider(memoryCache), new RelativeTtl(cacheDuration), new DefaultCacheKeyStrategy(),
    onCachePut: (context, key) => logger.LogInformation($"Caching '{key}'."),
    onCacheGet: (context, key) => logger.LogInformation($"Retrieving '{key}' from the cache."),
    onCachePutError: (context, key, exception) => logger.LogError(exception, $"Cannot add '{key}' to the cache."),
    onCacheGetError: (context, key, exception) => logger.LogError(exception, $"Cannot retrieve '{key}' from the cache."),
    onCacheMiss: (context, key) => logger.LogWarning($"Cache miss for '{key}'."));
var retryPolicy = Policy
    .Handle<HttpRequestException>()
    .WaitAndRetryAsync(retries, retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)),
        (exception, timeSpan, retry, context) =>
        {
            logger.LogWarning((HttpRequestException)exception, $"Cannot load external data from '{context.ExecutionKey}'. Retry {retry}/{retries}.");
        }
    );

policyRegistry.Add("CachedHttpResiliencePolicy", Policy.WrapAsync(cachePolicy, retryPolicy));
```

What I would expect it to do, is to load the data from cache (via `cachePolicy`), if it's available, or use the `retryPolicy` to retrieve the data on cache-miss/error. What actually happens is that the cache policy is never used and the data is always retrieved from the remote location, using the `retryPolicy`.

If I change the `CachedHttpResiliencePolicy` to contain only `cachePolicy`, instead of `Policy.WrapAsync(cachePolicy, retryPolicy)`, the the caching is working as expected. Anyhow, I'm losing the retry functionality.

Am I doing something incorrectly?

Thank you!

**Comment by reisenberger** [comment]
Good morning @TDabasinskas .  Could you post also the code where you place the call through the `PolicyWrap`?  If you are able to combine all into a [Minimal, Complete, and Verifiable example](https://stackoverflow.com/help/mcve) (even if just with a stub response method), that will also speed investigation.

For an immediate suggestion: test also with declaring the cache and retry policies as strongly-typed to `HttpResponseMessage`?  IE `var cachePolicy = Policy.CacheAsync<HttpResponseMessage>(...)`; `var retryPolicy = Policy<HttpResponseMessage>.Handle<HttpRequestException>(...)`

Assuming that works, this may be an issue specific to [non-generic](https://github.com/App-vNext/Polly/wiki/Non-generic-and-generic-policies) `CachePolicy` mid non-generic `PolicyWrap`.

Thanks

**Comment by reisenberger** [comment]
@TDabasinskas I can reproduce the issue.  I confirm it is limited to non-generic `CachePolicy` within a non-generic `PolicyWrap`.

I can confirm that if you define your policies as [generic policies](https://github.com/App-vNext/Polly/wiki/Non-generic-and-generic-policies#generic-policies-policytresult), as mentioned in the preceding comment, all functions fine.

I have identified a fix for non-generic `CachePolicy` within non-generic `PolicyWrap` used with the generic _method_ overloads `.Execute/Async<TResult>(...)`. This should follow in the next few days.

**Comment by reisenberger** [comment]
@TDabasinskas This is fixed in v5.5.0, due to be released to nuget shortly.

**Comment by TDabasinskas** [comment]
Hi @reisenberger,

Sorry for not replying sooner. I've finally tested the new version today and it seems it indeed fixed the issue. Thank you!
